sys/src/9/amd64/memory.c

```diff
@@ -13,18 +13,101 @@
 #include "dat.h"
 #include "fns.h"
 
+#include "amd64.h"
+
 void
 meminit(void)
 {
-	extern void asmmeminit(void);
+	int cx = 0;
+
+	for(PAMap *m = pamap; m != nil; m = m->next){
+		DBG("meminit: addr %#P end %#P type %d size %P\n",
+			m->addr, m->addr+m->size,
+			m->type, m->size);
+		PTE pgattrs = PteP;
+		switch(m->type){
+		default:
+			DBG("(Skipping)\n");
+			continue;
+		case PamKTEXT:
+			pgattrs |= PteG;
+			break;
+		case PamDEV:
+			pgattrs |= PtePCD;
+		case PamMEMORY:
+		case PamKRDWR:
+			pgattrs |= PteRW;
+		case PamACPI:
+		case PamPRESERVE:
+		case PamRESERVED:
+		case PamKRDONLY:
+			pgattrs |= PteNX;
+		}
+		mmukphysmap(UINT2PTR(machp()->MMU.pml4->va), m->addr, pgattrs, m->size);
+
+		/*
+		 * Fill in conf data.
+		 */
+		if (m->type != PamMEMORY)
+			continue;
+		if(cx >= nelem(conf.mem))
+			continue;
+		uintmem lo = ROUNDUP(m->addr, PGSZ);
+		conf.mem[cx].base = lo;
+		uintmem hi = ROUNDDN(m->addr + m->size, PGSZ);
+		conf.mem[cx].npage = (hi - lo)/PGSZ;
+		conf.npage += conf.mem[cx].npage;
+		DBG("cm %d: addr %#llx npage %lu\n",
+			cx, conf.mem[cx].base, conf.mem[cx].npage);
+		cx++;
+	}
+	mmukflushtlb();
+
+	/*
+	 * Fill in more legacy conf data.
+	 * This is why I hate Plan 9.
```

Comment:
This is a jmk comment and someday I'd like to figure out how to remove it.

Reply:
Removing this will require a bit more further study, because plan9 made naive assumptions about the amount of virtual address space available as a proxy for allocatable RAM.